src/shared/useModuleListeners.ts

```diff
@@ -0,0 +1,127 @@
+/**
+ * Copyright 2024 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { useRef, useCallback } from 'react';
+import { NativeEventEmitter, Platform, type NativeModule } from 'react-native';
+
+type ListenerMap<T> = {
+  [K in keyof T]?: NonNullable<T[K]>[];
+};
+
+export const useModuleListeners = <
+  T extends { [K in keyof T]: ((...args: any[]) => void) | undefined },
+>(
+  dispatcher: NativeModule,
+  androidBridge: string,
+  eventTypes: Array<keyof T>,
+  eventTransformer?: <K extends keyof T>(
+    eventKey: K,
+    ...args: unknown[]
+  ) => unknown[]
+): {
+  addListeners: (listeners: Partial<T>) => void;
+  removeListeners: (listeners: Partial<T>) => void;
+  removeAllListeners: () => void;
+} => {
+  const listenersRef = useRef<ListenerMap<T>>({});
+  const eventEmitterRef = useRef<NativeEventEmitter | null>(null);
+
+  const getIOSEventEmitter = useCallback(() => {
+    if (!eventEmitterRef.current) {
+      eventEmitterRef.current = new NativeEventEmitter(dispatcher);
+    }
+    return eventEmitterRef.current;
+  }, [dispatcher]);
+
+  const updateListeners = useCallback(() => {
```

Comment:
For some of these new files, let's please start building our unit test coverage.